ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -56,81 +55,139 @@ import           Ouroboros.Network.Mux.Types
-    -> Peer ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
-
-  -- |
-  -- Pipelined client and a server with the correspnding codec.
-  --
-  PipelinedClientAndServer
-    :: forall ps (st :: ps) m a.
-       Natural
-    -> Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> PeerPipelined ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
+-- * enumeration of server applications: this application type is mostly useful
+--   tests.
+--
+-- * enumeration of both client and server applications, e.g. a full node
+--   serving downstream peers using server side of each protocol and getting
+--   updates from upstream peers using client side of each of the protocols.
+--
+data MuxApplication (appType :: AppType) ptcl m  where
+  MuxClientApplication
+    -- Client application; most simple application will be @'runPeer'@ or
+    -- @'runPipelinedPeer'@ supplied with a codec and a @'Peer'@ for each
+    -- @ptcl@.  But it allows to handle resources if just application of
+    -- @'runPeer'@ is not enough.  It will be run as @'muxInitiator'@.
+    :: (ptcl -> Channel m ByteString ->  m ())
+    -> MuxApplication ClientApp ptcl m
+
+  MuxServerApplication
+    -- Server application; similarly to the @'MuxClientApplication'@ but it
+    -- will be run using @'muxResponder'@.
+    :: (ptcl -> Channel m ByteString ->  m ())
+    -> MuxApplication ServerApp ptcl m
+
+  MuxClientAndServerApplication
+    -- Client and server applications.
+    :: (ptcl -> Channel m ByteString ->  m ())
+    -> (ptcl -> Channel m ByteString ->  m ())
+    -> MuxApplication ClientAndServerApp ptcl m
 
+-- |
+-- Accessor for the client side of a @'MuxApplication'@.
+--
+clientApplication
+  :: ( Functor m
+     , HasClient appType ~ True
+     )
+  => (MuxApplication appType ptcl m)
+  -> (ptcl -> Channel m ByteString ->  m ())
+clientApplication (MuxClientApplication app) = \ptcl channel -> app ptcl channel
+clientApplication (MuxClientAndServerApplication app _) = \ptcl channel -> app ptcl channel
 
 -- |
--- Public network interface for 'ouroboros-network'.
+-- Accessor for the client side of a @'MuxApplication'@.
 --
-data NetworkInterface ptcl addr m r = NetworkInterface {
-      -- |
-      -- Address of the node to run.  The node will bind to this address, and
-      -- listen for incoming connections.  Some bearers do not have a notion of
-      -- address.
-      --
-      nodeAddress      :: addr,
+serverApplication
+  :: ( Functor m
+     , HasServer appType ~ True
+     )
+  => (MuxApplication appType ptcl m)
+  -> (ptcl -> Channel m ByteString ->  m ())
+serverApplication (MuxServerApplication app) = app
+serverApplication (MuxClientAndServerApplication _ app) = app
+
+getApplications
+  :: MuxApplication (appType :: AppType) ptcl m
+  -> ( Maybe (ptcl -> Channel m ByteString -> m ())
+     , Maybe (ptcl -> Channel m ByteString -> m ())
+     )
```

Comment:
With [my suggestion](https://github.com/input-output-hk/ouroboros-network/pull/543#discussion_r286639576) this would be fixed: the type would be `MuxApplication ptcl failure m c s -> ptcl -> (Channel m ByteString -> m c, Channel m ByteString -> m s)`

Reply:
Yes, this function can be removed and we can pattern match on `MuxApplication` directly.